subprojects/file-watching/src/main/java/org/gradle/internal/watch/vfs/impl/DefaultFileSystemWatchingHandler.java

```diff
@@ -131,17 +127,16 @@ public void beforeBuildFinished(boolean watchingEnabled) {
                 logWatchingError(reasonForNotWatchingFiles, FILE_WATCHING_ERROR_MESSAGE_AT_END_OF_BUILD);
                 reasonForNotWatchingFiles = null;
             }
-            root.update(currentRoot -> {
-                SnapshotHierarchy newRoot = removeSymbolicLinks(currentRoot);
-                newRoot = handleWatcherRegistryEvents(newRoot, "for current build");
+            rootReference.update(root -> {
+                removeSymbolicLinks(root);
+                handleWatcherRegistryEvents(root, "for current build");
                 if (watchRegistry != null) {
-                    newRoot = withWatcherChangeErrorHandling(newRoot, () -> watchRegistry.getFileWatcherUpdater().buildFinished());
+                    withWatcherChangeErrorHandling(root, () -> watchRegistry.getFileWatcherUpdater().buildFinished());
                 }
-                printStatistics(newRoot, "retains", "till next build");
-                return newRoot;
```

Comment:
Why did you change this to be in-place modification instead of the immutable stuff we had? I think the original transform-like behavior was cleaner.

Reply:
I think the mutable stuff is better since then we always get notifications about changes to the root reference. And the client actually wants to mutate the root.